sceptre/resolvers/__init__.py

```diff
@@ -49,62 +72,293 @@ def resolve(self):
+        rather than directly assigns it to avoid bugs related to shared objects in memory.
+
+        :param stack: The stack to assign the value to
+        :param value: The value to assign
+        """
+        cloned = self._clone_container_with_resolvers(value, stack)
+        setattr(stack, self.name, cloned)
+
+    def _clone_container_with_resolvers(
+        self,
+        container: T_Container,
+        stack: 'stack.Stack'
+    ) -> T_Container:
+        """Recurses into the container, cloning and setting up resolvers and creating a copy of all
+        nested containers.
+
+        :param container: The container being recursed into and cloned
+        :param stack: The stack the container is being copied for
+        :return: The fully copied container with resolvers fully set up.
+        """
+        def recurse(obj):
+            if isinstance(obj, Resolver):
+                return self.get_setup_resolver_for_stack(stack, obj)
+            if isinstance(obj, list):
+                return [
+                    recurse(item)
+                    for item in obj
+                ]
+            elif isinstance(obj, dict):
+                return {
+                    key: recurse(val)
+                    for key, val in obj.items()
+                }
+            return obj
 
-    class ResolveLater(object):
+        return recurse(container)
+
+    def _resolve_deferred_resolvers(self, stack: 'stack.Stack', container: T_Container):
+        def raise_if_not_resolved(attr, key, value):
+            # If this function has been hit, it means that after attempting to resolve all the
+            # ResolveLaters, there STILL are ResolveLaters left in the container. Rather than
+            # continuing to try to resolve (possibly infinitely), we'll raise a RecursiveGet to
+            # break that infinite loop. This situation would happen if a resolver accesses a resolver
+            # in the same container, which then accesses another resolver (possibly the same one) in
+            # the same container.
+            raise RecursiveResolve(f"Resolving Stack.{self.name[1:]} required resolving itself")
+
+        has_been_resolved_attr_name = f'{self.name}_is_resolved'
+        if not getattr(stack, has_been_resolved_attr_name, False):
+            # We set it first rather than after to avoid entering this block again on this property
+            # for this stack.
+            setattr(stack, has_been_resolved_attr_name, True)
+            _call_func_on_values(
+                lambda attr, key, value: value(),
+                container,
+                self.ResolveLater
+            )
+            # Search the container to see if there are any ResolveLaters left;
+            # Raise a RecursiveResolve if there are.
+            _call_func_on_values(
+                raise_if_not_resolved,
+                container,
+                self.ResolveLater
+            )
```

Comment:
This method makes it so we don't need to have that double-layer property on stacks for sceptre_user_data. It handles those scenarios where a resolver for sceptre_user_data might reference another property in sceptre_user_data.